### ISSUE TITLE
Formatted master incorrectly accepts blocks created before the format due to blockID(containerID) clash

### ISSUE THREAD

**Issue by jiacheliu3** [Alluxio/alluxio#14209]
**Alluxio Version:**
What version of Alluxio are you using?

**Describe the bug**
alluxio master will generate block container id from 0, after master format and restart without worker format, the old worker will report its block to alluxio master, not sure the master will accept the block by accident.

**To Reproduce**
Introduced in
https://github.com/Alluxio/alluxio/pull/14006#issuecomment-913334686

The new master has generated some blocks with container IDs starting from `0`. Then the worker from the previous cluster registers with old blocks (created with the old master, with container ID starting from `0`). If you are unlucky, a block on this old worker may have old blockID `0` and the new blockID `0` has been allocated to a totally irrelevant file. The new master, in this case, will mistakenly think this blockID is recognized and accepts this copy from the old worker. 

Then if the block lengths do not match (the new `0` vs old `0`), the master will throw the error referred in the link.

What's worse, if the block lengths DO match, the master will think this copy belongs to this totally irrelevant file. Then the Alluxio data is messed up without you noticing!

**Expected behavior**

**Urgency**
HIGH

**Additional context**

**Comment by jiacheliu3** [comment]
For the sake of completeness, I'm listing all possible combinations of scenarios, to make sure all the corner cases are scrutinized.

1. Formatted&restarted master + formatted&restarted worker
2. Formatted&restarted master + unformatted&restarted worker
3. Formatted&restarted master + unformatted&unrestarted worker
4. Unformatted&restarted master + formatted&restarted worker
5. Unformatted&restarted master + unformatted&restarted worker
6. Unformatted&restarted master + unformatted&unrestarted worker
7. Unformatted&unrestarted master + formatted&restarted worker
8. Unformatted&unrestarted master + unformatted&restarted worker
9. Unformatted&unrestarted master + unformatted&unrestarted worker

Also don't forget, we have the K8s environment where all paths are ephemeral unless you specify a persistent volume and mount it. On the worker we only have the following volumes. Adding more volumes adds operation overheads.
1. Volumes for tiered storage, as you need
2. Domain socket volume, to be shared with the alluxio client pod (within an application)
3. Log volume if you specifies

If we add a persistent state like `${ALLUXIO_HOME}`, we must not forget the possibility this is overlooked by the admin. We need to define what happens if the file is not copied during a worker update, and how this can be integrated into the existing scripts like `alluxio-start.sh`.

**Comment by jiacheliu3** [comment]
> @jiacheliu3 Just want to add a benefit for changing path scheme to include clusterId. It would be a must if we'd like to support running workers against multiple Alluxio clusters.
> 
> Other than that, I believe this is much preferable to starting block-Ids from arbitrary numbers. With this protocol, we'd also be sparing Alluxio master from flood of big registrations after format.

Agreed including clusterId into design will benefit supporting workers to be owned by multiple clusters.

On your second point, if you mean starting master with a manually specified container ID, it is really hard for the user to know what containerID to use. It will be better replaced by just giving the user an option when starting the worker to wipe out all tiered storage before starting the worker process, like embedding that into `alluxio-workers.sh`. These are equally manual.

**Comment by ggezer** [comment]
> > @jiacheliu3 Just want to add a benefit for changing path scheme to include clusterId. It would be a must if we'd like to support running workers against multiple Alluxio clusters.
> > Other than that, I believe this is much preferable to starting block-Ids from arbitrary numbers. With this protocol, we'd also be sparing Alluxio master from flood of big registrations after format.
> 
> Agreed including clusterId into design will benefit supporting workers to be owned by multiple clusters.
> 
> On your second point, if you mean starting master with a manually specified container ID, it is really hard for the user to know what containerID to use. It will be better replaced by just giving the user an option when starting the worker to wipe out all tiered storage before starting the worker process, like embedding that into `alluxio-workers.sh`. These are equally manual.
Let's not start block-container Ids from `System.currentTimeMillis()` is what I mean.

**Comment by xichen01** [comment]
> In our discussion today we discussed how to utilize the clusterID to determine whether a worker's blocks should be accepted or wiped out. The clusterID needs to be stored at the worker side and the master will compare that with the current clusterID, to determine whether the blocks on the worker are from the previous cluster (before the journal is formatted), or from the current cluster.
> 
> One problem is when the worker restarts, the worker cannot lose the clusterID and it cannot simple request the new clusterID from the master. If the worker loses the clusterID, there's no way the master can tell if the blocks on the worker are expired leftovers, or simply created in the new cluster. Therefore the worker has to persist the clusterID.
> 
> In terms of where to persist the clusterID, I prefer not to store it in the tiered storage. There are multiple tiers and directories and we should not make any of them special (special because if it holds the clusterID, we will rely on that). And if there are multiple copies of clusterID that will further introduce the problem of conflicts. I think we can rely on the `${ALLUXIO_HOME}` on the worker to always exist and be unique. That place should be okay if we need to persist some states for the worker.
> 
> The drawback of using `${ALLUXIO_HOME}/clusterID` is also obvious: In a container that ${ALLUXIO_HOME} directory is not persistent. You will have to create a volume there, like mounting a hostPath file to `${ALLUXIO_HOME}/clusterID`.
> 
> @apc999 @maobaolong @ggezer @yyongycy What do you guys think?

If the worker process does not have write permission to the ${ALLUXIO_HOME}  directory (such as we have a K8s environment and readonly mounts for ${ALLUXIO_HOME} ). in this case, we can not store the clusterld on any path except for the tiered storage. 
Is it a good idea to store clusterId to the environment variables? In this way the worker can store the clusterId in the environment of the system where it doesn't have write permission. Combine with previous ideas, we can try to save the clusterId to the ${ALLUXIO_HOME} directory, if failed store it in the environment.

**Comment by apc999** [comment]
@jiacheliu3  what's the status of this issue?

**Comment by jiacheliu3** [comment]
@apc999 The solution PR is still in design doc phase https://github.com/Alluxio/alluxio/pull/14258 under @xichen01

**Comment by xichen01** [comment]
@apc999 I will start coding on this PR recently [#14258](https://github.com/Alluxio/alluxio/pull/14258)
@jiacheliu3  I remember that most of the review of the design doc has been completed, do you have any suggestions?

**Comment by jiacheliu3** [comment]
> @apc999 I will start coding on this PR recently [#14258](https://github.com/Alluxio/alluxio/pull/14258) @jiacheliu3 I remember that most of the review of the design doc has been completed, do you have any suggestions?

The design doc is in a much better shape now! @ZhuTopher  and I will give it another review very soon. Thanks for the work!

**Comment by HelloHorizon** [comment]
@jiacheliu3 what is the current status of this issue?

**Comment by ZhuTopher** [comment]
@HelloHorizon I re-reviewed the design doc on Dec. 14, 2021. Not sure if the doc has been changed since then or not. I believe Jiacheng wanted to take another look at it still?

I haven't re-reviewed this PR itself since my initial pass either.

**Comment by jiacheliu3** [comment]
@HelloHorizon @ZhuTopher  The design doc has passed review and #14258 is in progress.
